src/Domain.Model/AggregateRoot.cs

```diff
@@ -0,0 +1,48 @@
+﻿namespace PetProjects.MtsManagementApi.Domain.Model
+{
+    using System;
+    using System.Collections.Generic;
+    using Infrastructure.CrossCutting.CQRS;
+
+    public abstract class AggregateRoot
+    {
+        private readonly List<Event> changes = new List<Event>();
+
+        public abstract Guid Id { get; }
+
+        public int Version { get; internal set; }
+
+        public IEnumerable<Event> GetUncommitedChanges()
+        {
+            return this.changes;
+        }
+
+        public void MarkChangesAsCommited()
+        {
+            this.changes.Clear();
+        }
+
+        public void LoadFromHistory(IEnumerable<Event> history)
+        {
+            foreach (var @event in history)
+            {
+                this.ApplyChange(@event, false);
+            }
+        }
+
+        protected void ApplyChange(Event @event)
+        {
+            this.ApplyChange(@event, true);
+        }
+
+        private void ApplyChange(Event @event, bool isNew)
+        {
+            this.AsDynamic().Apply(@event);
```

Comment:
It isn't clear where it is defined. But it must be defined inside de concrete Aggregate.
An example would be:

```
private void Apply(ItemCreated e)
{
    // state transition
   discountValue = e.price * discountPercentage;
}
```

What happens in here is that ApplyMethod converts itself into a dynamic type. Which allows the private method to be called and apply the change.

You can see a detailed explanation in the following link.
http://danielwhittaker.me/2014/11/15/aggregate-root-cqrs-event-sourcing/


I will add a comment in order to be clearer.

Reply:
So this Apply(event) method must be implemented in the concrete aggregate type otherwise this abstract class will not work correctly? And because it's reflection, failing to implement this method will only generate a error during runtime...
In the example that you described, would that actually work? The Event object would be automatically converted to an object of type ItemCreated if ItemCreated is a child class of Event?

Reply:
Yes, it will only generate an error during runtime, which is a bad thing...
ItemCreate must be a child class of Event, I will try to make it an interface instead of a class.
Yes, in this moment if the private method isn't implemented it will most certainly fail.

I will make changes to this and try to improve it, in order to not depend on reflection.

Maybe create an abstract method called `void Apply(Event event)` will work. Maybe it will substitute the reflection.